bastionzero/bzero
# Store split private keys in a per-agent configuration file

### This is one of those PRs that has no immediate effect, but supports other aspects of the solution and will be easier to review and merge in isolation

## Description of the change

As part of the [passwordless DB architecture](https://docs.google.com/document/d/1leYc0uJvF8oIge-G_x2702cmdoA61fkqvFbMFqDvccc/edit#heading=h.7topsnnuvkyh), we need a way for agents to store mappings between key shards and the database targets to which they authenticate. To support this, the agent will use a new configuration object with this structure:
```json
/* oldest key (toy example) */
[{"key":{"associatedPublicKey":{"n":null,"e":0},"d":null,"e":null},"targetIds":["62e7cbbe-f730-4aed-bb77-6f06749066be"]},
/* more recent key (toy example) */
{"key":{"associatedPublicKey":{"n":"2RxHnc7Yo7bAZbSNWrd4Qf1tXWLT0qPBQMNJiOcQdXkw9oRvjcD4LiBEtl3C4mjych/5s1OGwDCV5CNqeZMniqGL53vyEiRGfqAZes+L+1HmlYITzxAhFIISqNraWpTCVpKiSXV9Kd1+tLP7fJrviWPtPg1c86XR1MLdowEfk0xN5V0hc2ZRZqLgDlLCtLOlN3zD8AZF0lHyaVkbbBmsawej1y99o8fJlH56lmFcB3EB4HpQ9D0adg5R+qhH5A/mhevgISdsg+PHTzeGFG7tPRIWOc7b6sVZyXn8kswQcpaXosU8cVyCH91BZXGUEc8HV2Rtnglw1mXBE98uhevHOxGXD/0aA8nM2GnPI2Bb5l3YBTg4Iolt0EFqN52rC01sKqQDtLP18bE5pTrae+BCLzP3QKCl8fYCJdOqNK/9hN4BkDW+a78jdH9o1BB0WP4H+4kW6N20YDV9Z+/63ICl6JH0cSgAl4iEtukzqZKfxb2v5z1q9i7JQZbkc/ZmoIzsRBqv8QCd4DnTuUd4LhZftRGWdT6RKvxDsUFVceU5VK7qfjX/C+7fJuY1MmGI4KlegDh9yhut25LCaXO3In6FBravWuLKD9RDB/A/o9wgG4ZykqSQcvaZnU1yU6U3uWXMUu4KyhZU0G3yAKAGd6k+o9qwdPq6N/4znvp6jbq7n6c=","e":65537},"d":"MTIz","e":"NDU="},"targetIds":["62e7cbbe-f730-4aed-bb77-6f06749066be"]}]
```

When a new key is added, it goes at the end of the list. When a user is trying to authenticate to a virtual target, the agent will select the most recent key available that maps to that target. 

**For the customer beta**, we will support customers updating agents manually using a new `bzero` command, but this is not required for the demo-ready alpha.

### Some other notes:

  1. Fixed a broken test in the agentconfig test suite
  2. Removed a ginkgo import from a non-test file that was messing up our help options
  3. Tweaked how the systemd config client was using locks, and added a convenience wrapper for the `AcquireLock` flow so that we don't have to be so verbose
  4. **If the update to go1.18 goes through, I can simplify this code a bit** so that we don't need separate methods for fetching `KeyShardConfig` and `AgentConfig` data

## Testing

  1. Check out the `feat/generate-cert` branches in ZLI and backend (this will require a database migration and redeploying dev)
  5. Make sure you have a database target set up using your agent as a proxy
  6. Run `zlil generate certificate --all`
  7. (You should have a cert successfully returned to you)
  8. But more importantly, connect to the agent machine as root and check `cat /etc/bzero/bzero-user-keys.yaml`
  9. You should see the key shard you just generated in `/etc/bzero/keyshards.json`


**backend branch:** develop
**zli branch:** develop
**charts branch:** 

#### Ready to run system tests?


## Relevant release note information

Release Notes: Store split private keys in a per-agent configuration file

## Related JIRA tickets

Relates to JIRA: CWC-2130

## Have you considered the security impacts?

Does this PR have any security impact?


If yes, please explain:
